finality-aleph/src/finalization.rs

```diff
@@ -77,193 +80,99 @@ where
-                    .unwrap();
-                // Add a dummy extrinsic to make the block distinct from the one on chain
-                builder
-                    .push(Extrinsic::AuthoritiesChange(Vec::new()))
-                    .unwrap();
-                let block = builder.build().unwrap().block;
-                let hash = block.header.hash();
-                futures::executor::block_on(client.import(BlockOrigin::Own, block)).unwrap();
-                hash
-            })
-            .collect();
 
-        for i in 0..=n {
-            for j in 0..=n {
-                if i != j {
-                    let maybe_data = should_finalize(
-                        extra_children[i],
-                        AlephData::new(extra_children[j], j as u64),
-                        client.as_ref(),
-                        100u64,
-                    );
-                    assert!(maybe_data.is_none());
+            // At this point we know that the validation of the proposal went through, which means that the corresponding
+            // blocks are all available and thus the `proposal` is honest and correct. Note that there might be cases in which
+            // some other node got `false` from `is_proposal_available_and_correct`, in which case the othe node would not even
+            // reach this place. This is still fine, because it necessarily means that the other node must have marked the
+            // `proposal.top_block()` as a hopeless fork, and consequently, in the code below, we will return `None.
+
+            // WARNING: If we ever enable pruning, this code (and the code in Data Store) must be carefully analyzed
+            // for possible safety violations.
+
+            if proposal.number_top_block() <= last_finalized.num {
+                None
+            } else if let Some(block) = proposal.block_at_num(last_finalized.num) {
+                // We must check if the last_finalized block agrees with the one in the proposed branch
+                if block.hash == last_finalized.hash {
+                    Some(proposal.top_block())
+                } else {
+                    None
                 }
-            }
-        }
-    }
-
-    #[test]
-    fn should_finalize_for_incorrect_aleph_data() {
-        let mut client = Arc::new(TestClientBuilder::new().build());
-
-        let n = 5;
-        let blocks = create_chain(&mut client, n as u64);
-
-        for i in 0..n {
-            for j in i..n {
-                let maybe_data = should_finalize(
-                    blocks[i],
-                    AlephData::new(blocks[j], (j + 1) as u64),
-                    client.as_ref(),
-                    100u64,
-                );
-                assert!(maybe_data.is_none());
+            } else {
+                // The last_finalized block is strictly below the branch bottom block. We apply the proposal,
+                // as it agrees with last_finalized.
```

Comment:
Strictly speaking we should only finalize the top block if the last finalized block is just below the proposal branch (as opposed to arbitrarily far behind), which we haven't checked in this code. It has been checked in data availability, so this is not wrong, and this comment requires no changes, but I'm writing it here as I think it is a good argument for some of the changes I am proposing – with them this subtlety would not be relevant.

Reply:
Yes, I'm working on that :)